src/imports.rs

```diff
@@ -190,14 +190,18 @@ pub(crate) fn merge_use_trees(use_trees: Vec<UseTree>, merge_by: SharedPrefix) -
             continue;
         }
 
-        for flattened in use_tree.flatten() {
+        for mut flattened in use_tree.flatten() {
             if let Some(tree) = result
                 .iter_mut()
                 .find(|tree| tree.share_prefix(&flattened, merge_by))
             {
                 tree.merge(&flattened, merge_by);
             } else {
-                result.push(flattened);
+                // If this is the first tree with this prefix, handle potential trailing ::self
+                if merge_by == SharedPrefix::Module {
+                    flattened = flattened.nest_trailing_self();
+                }
+                result.push(flattened.nest_trailing_self());
```

Comment:
Is it necessary to make this nest call in the non-module case, and twice for the module case?

If so could you add or point to a test case that demonstrates why?

Reply:
Nope, that's definitely unnecessary, good spot. I added the conditional `merge_by` check later, I just forgot to remove the unconditional case.

Squashed and updated, ready for review again.